### PR TITLE
Diversity 2.4GHz (SX1280) Rx... aka Medley

### PR DESCRIPTION
**Pull request by JyeSmith**
This PR not only adds sx1280 diversity radios, but also has a LOT of other changes to the sx1280 lib I found a long the way.  Hopefully Ill remember them all and describe below.

### Diversity

Each radio can be called individually or together with the enum SX1280_Radio_Number_t.  Multiple radios can and/or together to group SPI transactions.  The idea is to make life easier if this is ever expanded to 3-4 radios.

```
typedef enum
{
    SX1280_Radio_1 = 1 << 0,
    SX1280_Radio_2 = 1 << 1,
    SX1280_Radio_All = 0xFF,
} SX1280_Radio_Number_t;
```

When a packet is received it is processes as normal.  If successful then all radio irqStatus are cleared.  Subsequent interrupts for the same packet will then not be processed since that radio returns a blank GetIrqStatus().

If a packets is processed and fails e.g. bad sync word or crc, then only that radios irqStatus is cleared.  Subsequent interrupts for the same OTA packet will then be processed as normal as its irqStatus has not been cleared.

Below is an trace of radio #1 failing HW CRC, then radio #2 having a successful packet processed.

image

~~Successful packet track is done by the variable `gotPacketThisInterval`.  Ideally instance->RXnbISR() would return true/false but I couldn't figure out how to do this.  Happy for someone to make a commit here and I will test it.~~ This has been done.

This PR supersedes https://github.com/ExpressLRS/ExpressLRS/pull/1304

### Other Changes

New target category of 'ELRS' to move away from 'DIY'.  These targets can of course be DIYed by users, but the majority of users will have commercial hardware and this naming will look more professional in Lua.  We should consider this for all new targets going into V3.

The FLRC CRC reg was accidentally being set to 0, resulting in no HW CRC check.  This has been fixed.

image

Added SX12XX_RX_SYNCWORD_ERROR to rx_status.  SX1280 packets will no longer be processed if the failed status is returned in irqStatus.

SPI transactions have been reordered to minimise latency.  Packet stats are only retrieved after a successfully being processed and RC data sent to the FC.  ClearIrqStatus is also now done after RC data is processed and sent.

WaitOnBusy() has been changed to replace NOP with micros().  Micros() takes for ever to called and adding it after digitalRead means it may not need to be called at all.  If it does need to be called then there is no need to waste time with a NOP, but instead waste time called micros().  There is a HUGE benefit here is reducing the delay and jitter in the first NSS call after an interrupt.  Compare the above trace to the below.  The delay between NSS going low and transaction beginning is gone.

image

Timing between the rx interrupt, and NSS going high after packet processing, is now a consistent 49us.  Before there was some jitter.

image

### CODE REVIEW

File: src/lib/SX1280Driver/SX1280_hal.h [hunk 1]
```diff
@@ -38,26 +38,31 @@ class SX1280Hal
     void end();
     void reset();
 
-    void ICACHE_RAM_ATTR WriteCommand(SX1280_RadioCommands_t command, uint8_t val, uint32_t busyDelay = 15);
-    void ICACHE_RAM_ATTR WriteCommand(SX1280_RadioCommands_t opcode, uint8_t *buffer, uint8_t size, uint32_t busyDelay = 15);
-    void ICACHE_RAM_ATTR WriteRegister(uint16_t address, uint8_t *buffer, uint8_t size);
-    void ICACHE_RAM_ATTR WriteRegister(uint16_t address, uint8_t value);
+    void NssHigh(SX1280_Radio_Number_t radioNumber);
+    void NssLow(SX1280_Radio_Number_t radioNumber);
```
Comment (JyeSmith) [inline src/lib/SX1280Driver/SX1280_hal.h:42]: Requires ICACHE_RAM_ATTR

File: src/lib/SX1280Driver/SX1280.cpp [hunk 2]
```diff
@@ -392,23 +399,32 @@ void ICACHE_RAM_ATTR SX1280Driver::TXnb()
...
-        SetMode(SX1280_MODE_FS);
+        SetMode(SX1280_MODE_FS, SX1280_Radio_All);
         TXnbISR();
         return;
     }
+
+    if (GPIO_PIN_NSS_2 != UNDEF_PIN)
+    {
+        if (lastSuccessfulPacketRadio == SX1280_Radio_1) instance->SetMode(SX1280_MODE_FS, SX1280_Radio_2); // Make sure the unused radio is in FS mode and will not receive the tx packet.
+        if (lastSuccessfulPacketRadio == SX1280_Radio_2) instance->SetMode(SX1280_MODE_FS, SX1280_Radio_1);
```
Comment (JyeSmith) [inline src/lib/SX1280Driver/SX1280.cpp:410]: Change to `SX1280_Radio_All - SX1280_Radio_2`.  If 3 or 4 radios are added, then all non used can be set to SX1280_MODE_FS in a single SPI transfer.

Comment (JyeSmith) [reply src/lib/SX1280Driver/SX1280.cpp:410]: Not possible yet with out reworking WaitOnBusy.

File: src/html/hardware.html [hunk 3]
```diff
@@ -49,6 +49,9 @@ <h1>Welcome to the <b>ExpressLRS</b> target configuration page<br />
 					<tr><td></td><td>NSS pin</td><td><input size='3' id='radio_nss' name='radio_nss' type='text'/></td></tr>
 					<tr><td></td><td>RST pin</td><td><input size='3' id='radio_rst' name='radio_rst' type='text'/></td></tr>
 					<tr><td></td><td>SCK pin</td><td><input size='3' id='radio_sck' name='radio_sck' type='text'/></td></tr>
+					<tr><td></td><td>BUSY_2 pin</td><td><input size='3' id='radio_busy_2' name='radio_busy_2' type='text'/></td>Busy pin for second SX1280</tr>
+					<tr><td></td><td>DIO1_2 pin</td><td><input size='3' id='radio_dio1_2' name='radio_dio1_2' type='text'/></td>Interrupt pin for seconds SX1280</tr>
```
Comment (pkendall64) [inline src/html/hardware.html:53]: i misspelled "second"

File: src/lib/SX1280Driver/SX1280_hal.cpp [hunk 4]
```diff
@@ -91,7 +109,23 @@ void SX1280Hal::init()
...
+void ICACHE_RAM_ATTR SX1280Hal::NssHigh(uint8_t radioNumber)
+{
+    if (radioNumber & SX1280_Radio_1) digitalWrite(GPIO_PIN_NSS, HIGH);
+    if (GPIO_PIN_NSS_2 != UNDEF_PIN && radioNumber & SX1280_Radio_2) digitalWrite(GPIO_PIN_NSS_2, HIGH);
+}
+
+void ICACHE_RAM_ATTR SX1280Hal::NssLow(uint8_t radioNumber)
+{
+    if (radioNumber & SX1280_Radio_1) digitalWrite(GPIO_PIN_NSS, LOW);
+    if (GPIO_PIN_NSS_2 != UNDEF_PIN && radioNumber & SX1280_Radio_2) digitalWrite(GPIO_PIN_NSS_2, LOW);
```
Comment (pkendall64) [inline src/lib/SX1280Driver/SX1280_hal.cpp:128]: Maybe just a single function `SetNSS` with 2 params?

**Review by pkendall64 — APPROVED** [review]

**Comment by pkendall64** [comment]
My changes are only niceties. I have tested the branch on regular hardware i.e. not tested the dual RX at all.

File: src/html/hardware.html (hunk 3 above)
Comment (JyeSmith) [reply src/html/hardware.html]: Dnoe

File: src/lib/SX1280Driver/SX1280_hal.cpp (hunk 4 above)
Comment (JyeSmith) [reply src/lib/SX1280Driver/SX1280_hal.cpp]: Added

**Comment by JyeSmith** [comment]
@pkendall64 I have rebased after Frank was merged and added some new generic targets.  Another once over might be required.

**Comment by JyeSmith** [comment]
This schematic is a reference for anyone looking to build the `Generic 2400 Diversity.json` or `Generic 2400 Diversity and VTx.json` targets.

Schematic_EpressLRS Diversity VTx Receiver_2022-06-05

**Review by AlessandroAU — APPROVED** [review]

**Comment by Martivip** [comment]
Hi @brandonrc any progress with Atto Duo target ?

**Pull request merged**